crates/circuit/src/dag_circuit.rs

```diff
@@ -5263,6 +5257,123 @@ impl DAGCircuit {
         Ok(new_node)
     }
 
+    /// Apply a [PackedOperation] to the back of the circuit.
+    #[allow(clippy::too_many_arguments)]
+    pub fn apply_operation_back(
+        &mut self,
+        py: Python,
+        op: PackedOperation,
+        qargs: &[Qubit],
+        cargs: &[Clbit],
+        params: Option<SmallVec<[Param; 3]>>,
+        extra_attrs: Option<Box<ExtraInstructionAttributes>>,
+        #[cfg(feature = "cache_pygates")] py_op: Option<PyObject>,
+        check: bool,
+    ) -> PyResult<NodeIndex> {
+        self.inner_apply_op(
+            py,
+            op,
+            qargs,
+            cargs,
+            params,
+            extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            py_op,
+            check,
+            false,
+        )
+    }
+
+    /// Apply a [PackedOperation] to the front of the circuit.
+    #[allow(clippy::too_many_arguments)]
+    pub fn apply_operation_front(
+        &mut self,
+        py: Python,
+        op: PackedOperation,
+        qargs: &[Qubit],
+        cargs: &[Clbit],
+        params: Option<SmallVec<[Param; 3]>>,
+        extra_attrs: Option<Box<ExtraInstructionAttributes>>,
+        #[cfg(feature = "cache_pygates")] py_op: Option<PyObject>,
+        check: bool,
+    ) -> PyResult<NodeIndex> {
+        self.inner_apply_op(
+            py,
+            op,
+            qargs,
+            cargs,
+            params,
+            extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            py_op,
+            check,
+            true,
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn inner_apply_op(
+        &mut self,
+        py: Python,
+        op: PackedOperation,
+        qargs: &[Qubit],
+        cargs: &[Clbit],
+        params: Option<SmallVec<[Param; 3]>>,
+        extra_attrs: Option<Box<ExtraInstructionAttributes>>,
+        #[cfg(feature = "cache_pygates")] py_op: Option<PyObject>,
+        check: bool,
+        front: bool,
```

Comment:
IMO, it'd make more sense to inline the contents of this method to the Rust `apply_operation_{front,back}` methods as well.

The most complex bit looks to be the bounds checking for the bits, which you should be able to simplify (see comment below). And, instead of passing `check` and `front` as a booleans here, we can avoid some indirection by leaving it up to the caller to actually do the checks and push it onto the DAG. This would save a conditional check, since the caller already knows whether to push onto the front or back. 

With all that gone, there's not much else going on here in terms of code de-duplication (there's no getting around the wad of parameters we need to pass through up to where we make the final `PackedInstruction`).